server/src/main/java/org/elasticsearch/repositories/blobstore/ShardSnapshotTaskRunner.java

```diff
@@ -38,23 +39,41 @@ public class ShardSnapshotTaskRunner {
     private final CheckedBiConsumer<SnapshotShardContext, FileInfo, IOException> fileSnapshotter;
 
     abstract static class SnapshotTask extends AbstractRunnable implements Comparable<SnapshotTask> {
-
-        private static final Comparator<SnapshotTask> COMPARATOR = Comparator.comparingLong(
-            (SnapshotTask t) -> t.context().snapshotStartTime()
-        ).thenComparing(t -> t.context().snapshotId().getUUID()).thenComparingInt(SnapshotTask::priority);
-
         protected final SnapshotShardContext context;
 
         SnapshotTask(SnapshotShardContext context) {
             this.context = context;
         }
 
+        public final SnapshotShardContext context() {
+            return context;
+        }
+
+        private long getSnapshotStartTime() {
+            return context().snapshotStartTime();
+        }
+
+        private String snapshotUUID() {
+            return context().snapshotId().getUUID();
+        }
+
         public abstract int priority();
 
-        public SnapshotShardContext context() {
-            return context;
+        @SuppressWarnings("resource")
+        private ShardId shardId() {
+            return context().store().shardId();
         }
 
+        private static final Comparator<SnapshotTask> COMPARATOR = Comparator
+            // Prefer work from the oldest running snapshot ...
+            .comparingLong(SnapshotTask::getSnapshotStartTime)
+            // ... tiebreaking on UUID just in case two of them start at the same millisecond ...
+            .thenComparing(SnapshotTask::snapshotUUID)
+            // ... then prefer starting new shard snapshots over uploading files for ongoing ones (for fast noop shard snapshots) ...
```

Comment:
Could you point me at the code where this happens? And is it the case that starting the shard snapshots AFTER finishing uploading files won't be noop shard snapshots?

I'm mostly thinking about the cases where later snapshots complete shard snapshots even if the older snapshots aren't done. It'd be nice to know for sure that those later ones aren't ever doing work.

Reply:
> Could you point me at the code where this happens?

What happens exactly? The code that prefers to start new shard snapshots over uploading files is the line that sorts by `SnapshotTask::priority`.

> I'm mostly thinking about the cases where later snapshots complete shard snapshots even if the older snapshots aren't done. It'd be nice to know for sure that those later ones aren't ever doing work.

I mean they could be doing work occasionally, if e.g. there are 10 threads and 9 of them are doing the last 9 tasks for an older snapshot, the free thread will start on the next task in the queue for the newer snapshot.

Reply:
Sorry, I was burying the purpose of my questions. The point I was trying to get at is whether prioritizing new shard snapshots is best, instead of prioritizing file uploads and the completion of existing shard snapshots.

Reply:
It doesn't seem obvious that the next snapshot of a shard will be a noop, whenever it gets scheduled, sooner vs later, because it could already have been a while since the first snapshot of that shard started.

Reply:
I wonder if it would be better to prioritize finishing file uploads of existing shard snapshots. Focus on getting one good overall snapshot, not spreading resources out to other snapshots.

Reply:
> I wonder if it would be better to prioritize finishing file uploads of existing shard snapshots.

AFAICT it doesn't matter, or at least there are fairly balanced pros and cons either way. The `ShardSnapshotTask` tasks are much more lightweight than the `FileSnapshotTask` ones so they don't meaningfully add to the WIP we're interested in limiting in this PR; meanwhile it'll save work overall to run them sooner.

Reply:
It could be thousands of shards in a new snapshot request that we prioritize starting up instead of running file uploads for an older snapshot.

We don't have any stats on this edge case, though, and the behavior in unchanged by this patch, so seems fine to go forward with 👍